packages/machine/src/protocol/uninstall-virtual-app.ts

```diff
@@ -5,87 +5,273 @@ import {
   ETHBucketAppState,
   NetworkContext
 } from "@counterfactual/types";
-import { AddressZero } from "ethers/constants";
 
 import { ProtocolExecutionFlow, StateChannel } from "..";
 import { Opcode } from "../enums";
 import { Context, ProtocolMessage, UninstallVirtualAppParams } from "../types";
+import { virtualChannelKey } from "../virtual-app-key";
+
+import { getChannelFromCounterparty } from "./utils/get-channel-from-counterparty";
 
 export const UNINSTALL_VIRTUAL_APP_PROTOCOL: ProtocolExecutionFlow = {
   0: [
-    p1,
+    addVirtualAppStateTransitionToContext,
+
     Opcode.OP_SIGN,
+
+    (message: ProtocolMessage, context: Context) => {
+      const {
+        intermediaryAddress,
+        initiatingAddress
+      } = message.params as UninstallVirtualAppParams;
+
+      context.outbox[0] = {
+        ...message,
+        seq: 1,
+        fromAddress: initiatingAddress,
+        toAddress: intermediaryAddress,
+        signature: context.signatures[0]
+      };
+    },
+
     // send M1, wait for M4
     Opcode.IO_SEND_AND_WAIT,
 
-    uninstallLeftAgreement,
+    // TODO: Add signature verification
```

Comment:
Adding signature verification is not a part of this PR?

Reply:
I'm supposed to add it in a separate PR